src/couch_httpd_csrf.erl

```diff
@@ -0,0 +1,147 @@
+        {Csrf, Csrf} ->
+            %% cookie and header match, but is it valid?
+            ok = validate(Csrf),
+
+            %% check expiration
+            {_, Timestamp, _} = decode_cookie(Csrf),
+            MaxAge = max_age(),
+            case Timestamp + MaxAge < timestamp() of
+                true ->
+                    throw({forbidden, <<"CSRF Cookie expired">>});
+                false ->
+                    ok
+            end;
+        _ ->
+            throw({forbidden, <<"CSRF Cookie/Header mismatch">>})
+    end;
+%% Check that we generated this CSRF token
+validate(Csrf) when is_list(Csrf) ->
+    {Token, Timestamp, ActualHmac} = decode_cookie(Csrf),
+    Secret = ensure_csrf_secret(),
+    ExpectedHmac = crypto:sha_mac(Secret, <<Token/binary, Timestamp:32>>),
+    case couch_passwords:verify(ActualHmac, ExpectedHmac) andalso
+        timestamp() > Timestamp of
+        true ->
+            ok;
+        false ->
+            throw({forbidden, <<"Invalid CSRF value">>})
+    end.
+
+
+headers(#httpd{} = Req, Headers) ->
+    Header = couch_httpd:header_value(Req, "X-CouchDB-CSRF"),
+    case {csrf_from_req(Req), csrf_in_headers(Headers), Header} of
+        {undefined, false, "true"} ->
+            [make_csrf_cookie() | Headers];
+        {Csrf, false, Csrf} when Csrf /= undefined ->
+            {_, Timestamp, _} = decode_cookie(Csrf),
+            MaxAge = max_age(),
+            TimeLeft = Timestamp + MaxAge - timestamp(),
+            case TimeLeft < MaxAge * 0.5 of
+                true ->
+                    [make_csrf_cookie() | Headers];
+                false ->
+                    Headers
+            end;
+        _ ->
+	    Headers
+    end.
+
+
+make_csrf_cookie() ->
+    Secret = ?l2b(ensure_csrf_secret()),
+    Token = crypto:rand_bytes(8),
+    Timestamp = timestamp(),
+    Data = <<Token/binary, Timestamp:32>>,
+    Hmac = crypto:sha_mac(Secret, Data),
+    mochiweb_cookies:cookie("CouchDB-CSRF",
+        couch_util:encodeBase64Url(<<Data/binary, Hmac/binary>>),
+
+        [{path, "/"}, {max_age, max_age()}]).
+
+csrf_from_req(#httpd{} = Req) ->
+    case couch_httpd:header_value(Req, "Cookie") of
+	undefined ->
+	    undefined;
+	Value ->
+	    Cookies = mochiweb_cookies:parse_cookie(Value),
+	    couch_util:get_value("CouchDB-CSRF", Cookies)
+    end.
+
+
+csrf_in_headers(Headers) when is_list(Headers) ->
+    lists:any(fun is_csrf_header/1, Headers).
+
+
+is_csrf_header({"Set-Cookie", [$C, $o, $u, $c, $h, $D, $B, $-, $C, $S, $R, $F, $= | _]}) ->
+    true;
+is_csrf_header(_) ->
+    false.
+
+
+ensure_csrf_secret() ->
+    case config:get("couch_httpd_auth", "csrf_secret", undefined) of
+        undefined ->
+            NewSecret = ?b2l(couch_uuids:random()),
+            config:set("couch_httpd_auth", "csrf_secret", NewSecret),
+            NewSecret;
+        Secret -> Secret
+    end.
+
+
+decode_cookie(Cookie) ->
+    Cookie1 = couch_util:decodeBase64Url(Cookie),
```

Comment:
How about to catch a case if we receive malformed cookie here? While HTTP 500 because of badmatch or badarg is Erlang way, but more user friendly will be to be HTTP 400 here.